### PR TITLE
Add doTransform function for twists or wrenches

### PR DESCRIPTION
**Pull request by AndyZe**
I think this would be convenient for a lot of people!

It would be cool if there were a way to return an error if the VectorXd is not 6 long.

### CODE REVIEW

**Comment by AndyZe** [comment]
Pinging for review.

**Comment by aprotyas** [comment]
> It would be cool if there were a way to return an error if the VectorXd is not 6 long.

Naive thought, but you could check at runtime with `vector.size()` right?

File: tf2_eigen/include/tf2_eigen/tf2_eigen.hpp [hunk 1]
```diff
@@ -127,6 +127,51 @@ void doTransform(
...
+  twist_transform.block(0, 0, 3, 3) = affine_transform.rotation();
+  // upper right 3x3 block is all zeros
+  twist_transform.block(0, 3, 3, 3) = Eigen::MatrixXd::Zero(3, 3);
+  // lower left 3x3 block is tricky. See https://core.ac.uk/download/pdf/154240607.pdf
+  Eigen::MatrixXd pos_vector_3x3(3, 3);
+  // Disable formatting checks so the matrix remains human-readable
+/* *INDENT-OFF* */
+  pos_vector_3x3(0, 0) = 0;  pos_vector_3x3(0, 1) = -affine_transform.translation().z();  pos_vector_3x3(0, 2) = affine_transform.translation().y(); // NOLINT
+  pos_vector_3x3(1, 0) = affine_transform.translation().z();  pos_vector_3x3(1, 1) = 0;  pos_vector_3x3(1, 2) = -affine_transform.translation().x(); // NOLINT
+  pos_vector_3x3(2, 0) = -affine_transform.translation().y();  pos_vector_3x3(2, 1) = affine_transform.translation().x();  pos_vector_3x3(1, 2) = 0; // NOLINT
```
Comment (aprotyas) [inline tf2_eigen/include/tf2_eigen/tf2_eigen.hpp]: Just a thought: I like the comma initializer syntax Eigen provides. How readable that is, is subjective.

```suggestion
pos_vector_3x3 << 0, -affine_transform.translation().z(), affine_transform.translation().y(),
                  affine_transform.translation().z(), 0, -affine_transform.translation().x(),
                  -affine_transform.translation().y(), affine_transform.translation().x(), 0;
```

Comment (aprotyas) [reply tf2_eigen/include/tf2_eigen/tf2_eigen.hpp]: Ugh, the Github suggestion makes it look ugly. This is what I mean:
```cpp
pos_vector_3x3 << 0, -affine_transform.translation().z(), affine_transform.translation().y(),
                  affine_transform.translation().z(), 0, -affine_transform.translation().x(),
                  -affine_transform.translation().y(), affine_transform.translation().x(), 0;
```

**Review by aprotyas — COMMENTED** [review]
Math LGTM on a cursory pass. Will leave a proper review soon.
One style suggestion in-line.

File: tf2_eigen/include/tf2_eigen/tf2_eigen.hpp (hunk 1 above)
Comment (AndyZe) [reply tf2_eigen/include/tf2_eigen/tf2_eigen.hpp]: Thanks for the review! That suggestion looks much nicer.

File: tf2_eigen/include/tf2_eigen/tf2_eigen.hpp [hunk 2]
```diff
@@ -127,6 +127,51 @@ void doTransform(
...
+  // upper right 3x3 block is all zeros
+  twist_transform.block(0, 3, 3, 3) = Eigen::MatrixXd::Zero(3, 3);
+  // lower left 3x3 block is tricky. See https://core.ac.uk/download/pdf/154240607.pdf
+  Eigen::MatrixXd pos_vector_3x3(3, 3);
+  // Disable formatting checks so the matrix remains human-readable
+/* *INDENT-OFF* */
+pos_vector_3x3 << 0, -affine_transform.translation().z(), affine_transform.translation().y(),
+                  affine_transform.translation().z(), 0, -affine_transform.translation().x(),
+                  -affine_transform.translation().y(), affine_transform.translation().x(), 0;
+/* *INDENT-ON* */
```
Comment (clalancette) [inline tf2_eigen/include/tf2_eigen/tf2_eigen.hpp]:
```suggestion
  /* *INDENT-OFF* */
  pos_vector_3x3 << 0, -affine_transform.translation().z(), affine_transform.translation().y(),
                    affine_transform.translation().z(), 0, -affine_transform.translation().x(),
                    -affine_transform.translation().y(), affine_transform.translation().x(), 0;
  /* *INDENT-ON* */
```

**Review by clalancette — CHANGES_REQUESTED** [review]
One minor formatting fix, then I think we can run CI on this.

**Review by clalancette — APPROVED** [review]

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15347/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10058/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13027/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15549/)

**Comment by clalancette** [comment]
Failures on Windows CI are unrelated to this change.  But since Windows is the most likely to have problems, I'm going to wait until Windows is fixed and re-run it before merging.

**Comment by clalancette** [comment]
The warning on Windows is likely because this PR needs to be rebased.  @AndyZe mind doing that?

**Comment by aprotyas** [comment]
CI (build: `--packages-up-to tf2_eigen`, test: `--packages-select tf2_eigen`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15352/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10062/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13036/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15551/)

**Comment by clalancette** [comment]
The macOS failure is in the nightlies as well, so this looks good.  Thanks for iterating!

**Pull request merged**